lib/web/sign.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package web
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"crypto/x509/pkix"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/julienschmidt/httprouter"
+	"golang.org/x/exp/slices"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/client/identityfile"
+	"github.com/gravitational/teleport/lib/httplib"
+	"github.com/gravitational/teleport/lib/reversetunnel"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+/* signCertKeyPair returns the necessary files to set up mTLS for other services
+This is the equivalent of running the tctl command
+As an example, requesting:
+POST /webapi/sites/mycluster/sign
+{
+	"hostname": "pg.example.com",
+	"ttl": "2190h",
+	"format": "db"
+}
+
+Should be equivalent to running:
+   tctl auth sign --host=pg.example.com --ttl=2190h --format=db
+
+This endpoint returns a tar.gz compressed archive containing the required files to setup mTLS for the service.
+*/
+func (h *Handler) signCertKeyPair(w http.ResponseWriter, r *http.Request, p httprouter.Params, site reversetunnel.RemoteSite) (interface{}, error) {
+	ctx := r.Context()
+	req, err := parseSignCertKeyPair(r)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key, err := client.NewKey()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	subject := pkix.Name{CommonName: req.Hostname}
+	csr, err := tlsca.GenerateCertificateRequestPEM(subject, key.Priv)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clusterAPI := h.auth.proxyClient
+	resp, err := clusterAPI.GenerateDatabaseCert(
+		ctx,
+		&proto.DatabaseCertRequest{
+			CSR: csr,
+			// Important to include SANs since CommonName has been deprecated
+			ServerNames: []string{req.Hostname},
+			// Include legacy ServerName for compatibility.
+			ServerName:    req.Hostname,
+			TTL:           proto.Duration(req.TTL),
+			RequesterName: proto.DatabaseCertRequest_TCTL,
+		},
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
Can this logic be factored out of "tctl auth sign" implementation and reused in both places? For example, Mongo, Cockroach and Snowflake require some special handling, as it is it wouldn't generate secrets correctly for them.